cranelift/codegen/src/isa/aarch64/lower.isle

```diff
@@ -2374,6 +2374,76 @@
        (add_with_flags ty a b)
        (invalid_reg)))
 
+;;; Rules for `iadd_cout` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; For values smaller than a register, we do a normal `add` with both arguments
+;; sign extended. We then check if the output sign bit has flipped.
+(rule 0 (lower (has_type (fits_in_16 ty) (iadd_cout a b)))
+      (let ((extend ExtendOp (lower_extend_op ty $true))
+            
+            ;; Instead of emitting two `sxt{b,h}` we do one as an instruction and
+            ;; the other as an extend operation in the `add` instruction.
+            ;;
+            ;; sxtb    a_sext, a
+            ;; add     out, a_sext, b, sxtb
+            ;; cmp     out, out, sxtb
+            ;; cset    out_carry, ne
+            (a_sext Reg (put_in_reg_sext32 a))
+            (out Reg (add_extend_op ty a_sext b extend))
+            (out_carry Reg (with_flags_reg 
+                  (cmp_extend (OperandSize.Size32) out out extend)
+                  (cset (Cond.Ne)))))
+      (output_pair
+            (value_reg out)
+            (value_reg out_carry))))
+
+
+;; For register sized add's we just emit a adds+cset, without further masking.
+;;
+;; adds out, a, b
+;; cset carry, hs
+(rule 1 (lower (has_type (ty_32_or_64 ty) (iadd_cout a b)))
+      (let ((out ValueRegs
+              (with_flags
+                  (add_with_flags_paired ty a b)
+                  (cset_paired (Cond.Vs)))))
+      (output_pair
+            (value_regs_get out 0)
+            (value_regs_get out 1))))
+
+
+;; For i128 we just do a regular i128 add while also setting flags, and then
+;; `cset` the overflow output.
+;;
+;; adds out_lo, a_lo, b_lo
+;; adcs out_hi, b_hi, b_hi
+;; cset out_carry, vs
+(rule 2 (lower (has_type $I128 (iadd_cout x y)))
+      (let
+          ;; Get the high/low registers for `x`.
+          ((x_regs ValueRegs x)
+           (x_lo Reg (value_regs_get x_regs 0))
+           (x_hi Reg (value_regs_get x_regs 1))
+
+           ;; Get the high/low registers for `y`.
+           (y_regs ValueRegs y)
+           (y_lo Reg (value_regs_get y_regs 0))
+           (y_hi Reg (value_regs_get y_regs 1))
+           
+           (out_lo WritableReg (temp_writable_reg $I64))
+           (out_hi WritableReg (temp_writable_reg $I64))
+           (out_carry WritableReg (temp_writable_reg $I64))
+
+           ;; Do a i128 add which is a `adds` followed by an `adcs`, then emit
+           ;; a `cset` for the carry out.
+           (_ Unit (emit (MInst.AluRRR (ALUOp.AddS) (OperandSize.Size64) out_lo x_lo y_lo)))
+           (_ Unit (emit (MInst.AluRRR (ALUOp.AdcS) (OperandSize.Size64) out_hi x_hi y_hi)))
+           (_ Unit (emit (MInst.CSet out_carry (Cond.Vs)))))
```

Comment:
I couldn't find a way to do this with the current `with_flags` mechanism, it looks like it can only return at most two registers, but I need the result of all three.

Is there a better way of doing it?

Reply:
You should be able to define an `output_trio` extern constructor, the same way that `output_pair` is defined in `cranelift/codegen/src/machinst/isle.rs` and `cranelift/codegen/src/prelude_lower.isle`.

Reply:
Alternatively, if you don't want to deal with that at the moment, we could hold off on `i128` support for now.

But yeah, there should never be raw `emit` calls in `*/lower.isle`.